src/transform/deadCode.ml

```diff
@@ -0,0 +1,68 @@
+open Batteries
+open GoblintCil
+
+
+(** Filter statements out of a block (recursively). CFG fields (prev, next, etc.) are no longer valid after calling.
+    Returns true if anything is left in block, false if the block is now empty.
+    Invariants:
+    - f (goto label) ==> f (labelled stmt), i.e. if a goto statement is not filtered out, the target may not be filtered out either.
+    - block may not contain switch statements *)
+let filter_map_block f (block : Cil.block) : bool =
+  (* blocks and statements: modify in place, then return true if should be kept *)
+  let rec impl_block block =
+    block.bstmts <- List.filter impl_stmt block.bstmts;
+    block.bstmts <> []
+  and impl_stmt stmt =
+    if not (f stmt) then false
+    else
+      match stmt.skind with
+      | If (_, b1, b2, _, _) ->
+        (* be careful to not short-circuit, since call to impl_block b2 is always needed for side-effects *)
+        let keep_b1, keep_b2 = impl_block b1, impl_block b2 in keep_b1 || keep_b2
+      | Switch _ -> failwith "switch statements must be removed"
+        (* handling switch statements correctly would be very difficult; consider that the switch
+        labels may be located within arbitrarily nested statements within the switch statement's block *)
+      | Loop (b, _, _, _, _) ->
+        impl_block b
+      | Block b ->
+        impl_block b
+      | sk -> true
```

Comment:
This match should enumerate all constructors explicitly and be explicit about each construction and what is assumed to be transformed away. Like `CfgTools` does it.